test/gtest/uct/ib/test_sockaddr.cc

```diff
@@ -676,93 +733,149 @@ UCS_TEST_P(test_uct_cm_sockaddr, cm_server_reject)
                              client_connect_cb, client_disconnect_cb, this);
     }
 
     /* wait for the server to connect to all the clients */
     wait_for_client_server_counters(&m_server_connect_cb_cnt,
-                                    &m_client_connect_cb_cnt, num_clients);
+                                    &m_client_connect_cb_cnt, m_enhance_clients_num);
 
-    EXPECT_EQ(num_clients, m_server_recv_req_cnt);
-    EXPECT_EQ(num_clients, m_client_connect_cb_cnt);
-    EXPECT_EQ(num_clients, m_server_connect_cb_cnt);
-    EXPECT_EQ(num_clients, (int)m_server->num_eps());
+    EXPECT_EQ(m_enhance_clients_num, m_server_recv_req_cnt);
+    EXPECT_EQ(m_enhance_clients_num, m_client_connect_cb_cnt);
+    EXPECT_EQ(m_enhance_clients_num, m_server_connect_cb_cnt);
+    EXPECT_EQ(m_enhance_clients_num, (int)m_server->num_eps());
 
     /* Disconnect */
-    for (int i = 0; i < num_clients; ++i) {
+    srand(seed);
+    UCS_TEST_MESSAGE << "Using seed: " << seed;
+    pthread_mutex_init(&m_lock, NULL);
+    m_all_eps = (ep_state_t*)calloc(2 * m_enhance_clients_num, sizeof(ep_state_t));
+
+    /* save all the clients' and server's eps in the m_all_eps array */
+    for (i = 0; i < m_enhance_clients_num; ++i) {
         /* first 2 entities are m_server and m_client */
-        client_test = &m_entities.at(2 + i);
-        ASSERT_TRUE(client_test != m_client);
+        m_all_eps[i].ep    = m_entities.at(2 + i).ep(0);
+        m_all_eps[i].state = 0;
+        m_all_eps[m_enhance_clients_num + i].ep    = m_server->ep(i);
+        m_all_eps[m_enhance_clients_num + i].state = 0;
+    }
+
+    /* Disconnect */
+    /* go over the eps array and for each ep - use rand() to decide whether or
+     * not it should initiate a disconnect */
+    for (i = 0; i < (2 * m_enhance_clients_num); ++i) {
+        if (ucs::rand() % 2) {
+            /* don't start a disconnect on an ep that was already disconnected */
+            pthread_mutex_lock(&m_lock);
+            if (!(m_all_eps[i].state & TEST_CM_EP_STATE_DISCONNECT_CB_INVOKED)) {
+                m_all_eps[i].state |= TEST_CM_EP_STATE_DISCONNECT_INITIATOR;
+                ASSERT_UCS_OK(uct_ep_disconnect(m_all_eps[i].ep, 0));
+            }
+            pthread_mutex_unlock(&m_lock);
+        }
+    }
 
-        cm_disconnect(client_test);
+    /* wait for all the disconnect flows that began, to complete.
+     * if an ep initiated a disconnect, its disconnect callback should have been
+     * called, and so it the disconnect callback of its remote peer ep. */
+    i = 0;
+    deadline = ucs_get_time() + ucs_time_from_sec(10 * DEFAULT_TIMEOUT_SEC) *
+                                ucs::test_time_multiplier();
+start_over:
+    while ((i < (2 * m_enhance_clients_num)) && (ucs_get_time() < deadline)) {
+        if (m_all_eps[i].state == TEST_CM_EP_STATE_DISCONNECT_INITIATOR) {
+            progress();
+            i = 0;
```

Comment:
why need to reset counter?

Reply:
Because if we found one ep that initiated disconnect but its callback wasn't called yet, no point in checking the rest, can start going over the array again (after calling progress)

Reply:
I believe it may be simplified. For example, if you add a counter to the loop above, you can just wait like this:
```
while ((counter * 2) < (m_server_disconnect_cnt + m_client_disconnect_cnt)) {
    progress();
}
assert((counter * 2) == (m_server_disconnect_cnt + m_client_disconnect_cnt));
assert(m_server_disconnect_cnt == m_client_disconnect_cnt);
```

Reply:
how would this work? if, for example, all eps are selected for disconnect, so the counter would be equal to 200. m_server_disconnect_cnt and m_server_disconnect_cnt are expected to be equal to 100 each. so it wouldn't even go inside the loop. if i change (<) to (!=) and add time deadline, it will timeout.
why is what i have now not good?

Reply:
yep, sorry, need to replace "<" with ">"
number of invoked callbacks must be equal to "disconnect" calls, right? So, why it should timeout?
> why is what i have now not good?

it's non-obvious that exactly you are testing with many repeats on the array.

Reply:
> why is what i have now not good?

It requires deep reading to understand that test is doing. 
I see now why my previous approach won't work...
Let's count initiators in first loop then decrement in initiator EP's disconnect CB then wait counter == 0